capstone/src/test/java/com/google/sps/RecommenderAlgorithmTest.java

```diff
@@ -0,0 +1,149 @@
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.sps.service.DatabaseService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.HashMap;
+ 
+@RunWith(JUnit4.class)
+public final class RecommenderAlgorithmTest {
+    
+    private final LocalServiceTestHelper helper =
+        new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+    private static final List<String> TEST_CATEGORY_TAGS = Tag.CATEGORY_TAGS;
+
+    private static User TEST_USER_SMITH;
+    private static User TEST_USER_JOHN;
+    private static User TEST_USER_BILL;
+    private static User TEST_USER_SAM;
+    private static Room TEST_ROOM_POETRY;
+    private static Room TEST_ROOM_RUN;
+    private static Room TEST_ROOM_COOK; 
+    private static Room TEST_ROOM_COOK_TWO;
+
+    @Before
+    public void setUp() {
+        helper.setUp();
+        TEST_USER_SMITH = new User("1234", "Smith");
+        TEST_USER_JOHN = new User("2341", "John");
+        TEST_USER_BILL = new User("3412", "Bill");
+        TEST_USER_SAM = new User("4123", "Sam");
+        DatabaseService.save(TEST_USER_SMITH.getUserEntity());
+        DatabaseService.save(TEST_USER_JOHN.getUserEntity());
+        DatabaseService.save(TEST_USER_BILL.getUserEntity());
+        DatabaseService.save(TEST_USER_SAM.getUserEntity());
+ 
+        TEST_ROOM_POETRY = new Room(TEST_USER_SMITH, "How to Teach Poetry", "Learn how to teach Poetry");
+        TEST_ROOM_RUN = new Room(TEST_USER_JOHN, "Run Efficiently", "Improve your running experience");
+        TEST_ROOM_COOK = new Room(TEST_USER_BILL, "Iron Chef Master Course", "Cook Better");
+        TEST_ROOM_COOK_TWO = new Room(TEST_USER_BILL, "Iron Chef Master Course II", "Cook Better Part 2");
+        DatabaseService.save(TEST_ROOM_POETRY.getRoomEntity());
+        DatabaseService.save(TEST_ROOM_RUN.getRoomEntity());
+        DatabaseService.save(TEST_ROOM_COOK.getRoomEntity());
+        DatabaseService.save(TEST_ROOM_COOK_TWO.getRoomEntity());
+    }
+ 
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+ 
+    @Test
+    public void populateUserTagMapOneTagOneClassroomTest() { 
+        TEST_ROOM_RUN.addTag(Tag.TAG_FITNESS);
+        TEST_ROOM_RUN.addFollower(TEST_USER_SMITH);
+        DatabaseService.save(TEST_ROOM_RUN.getRoomEntity());
+
+        HashMap<String, Integer> userTagMap = RecommenderAlgorithm.populateUserTagMap(TEST_USER_SMITH);
+        for(String tag : TEST_CATEGORY_TAGS){
+            if(tag.equals(Tag.TAG_FITNESS)){
+                Assert.assertEquals((int) userTagMap.get(tag), 1);
+                continue;
```

Comment:
`else` preferred to `continue` + fallthrough since it is easier to track, and harder to accidentally make a bug out of (its pretty easy in a big nested if structure to think its OK to delete a `continue`, only to not realize that certain fallthroughs were critical). As a rule of thumb, `continue` is an antipattern (though not always conveniently avoidable)

Reply:
This makes more sense. I think I got to used to continue that I use it since it's "faster" in some cases since it doesn't finish the loop. Thanks for the tip!

Reply:
Its actually not faster since the compiler breaks both the "if" and the "else" part into control flow blocks the same as it would break the "up to the continue" and "after the continue" part into control flow blocks. In any case, if it is faster its on the scale of nanoseconds. Its nearly always (except in extreme performance critical scenarios) best to optimize for easiest to read and modify without making bugs since the compiler is pretty smart and will make better optimizations automatically than you can manually.